src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java

```diff
@@ -19,114 +19,96 @@
-          "[-extraAnalyzers path/to/analyzer3.jar [path/to/analyzer4.jar] ...]";
+import picocli.CommandLine;
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.ParameterException;
+import picocli.CommandLine.Parameters;
+import picocli.CommandLine.Spec;
+import picocli.CommandLine.Model.CommandSpec;
+
+@Command
+public class ServerMain implements Callable<Integer> {
 
   private final PrintStream out;
-  private final PrintStream err;
 
-  public ServerMain(PrintStream out, PrintStream err) {
+  @Parameters(index = "0", description = "The port to which sonarlint should connect to.", defaultValue = "-1")
+  private int deprecatedJsonRpcPort;
+
+  @Option(names = "-jsonRpcPort", description = "The port to which sonarlint should connect to.")
+  private Optional<Integer> jsonRpcPort;
+
+  @Option(names = "-stdio", description = "The actual transport channel will be stdio")
+  private boolean useStdio;
+
+  @Option(names = "-analyzers", arity = "1..*", description = "A list of paths to the analyzer JARs that should be used to analyze the code.")
+  private List<Path> analyzers = new ArrayList<>();
+
+  @Option(names = "-extraAnalyzers", arity = "1..*")
+  private List<Path> extraAnalyzers = new ArrayList<>();
+
+  @Spec
+  private CommandSpec spec;
+
+  public ServerMain(PrintStream out, PrintStream in) {
     this.out = out;
-    this.err = err;
   }
 
-  public static void main(String... args) {
-    new ServerMain(System.out, System.err).startLanguageServer(args);
+  public List<Path> getAnalyzers() {
+    return ImmutableList.copyOf(analyzers);
   }
 
-  static int getIndexOfNextParam(int start, String[] args) {
-    for (int i = start + 1; i < args.length; i++) {
-      if (args[i].startsWith("-")) {
-        return i;
-      }
-    }
-    return -1;
+  public List<Path> getExtraAnalyzers() {
+    return ImmutableList.copyOf(extraAnalyzers);
   }
 
-  public void startLanguageServer(String... args) {
-    if (args.length < 1) {
-      err.println(USAGE);
-      exitWithError();
-    }
-    var jsonRpcPort = parsePortArgument(args);
-
-    var analyzers = extractAnalyzers(args);
-    var extraAnalyzers = extractExtraAnalyzers(args);
-
-    out.println("Binding to " + jsonRpcPort);
-    try {
-      SonarLintLanguageServer.bySocket(jsonRpcPort, analyzers, extraAnalyzers);
-    } catch (IOException e) {
-      err.println("Unable to connect to the client");
-      e.printStackTrace(err);
-      exitWithError();
-    }
-  }
+  @Override
+  public Integer call() throws Exception {
+    validate();
 
-  Collection<Path> extractAnalyzers(String[] args) {
-    var indexOfAnalyzersParam = List.of(args).indexOf(ANALYZERS_KEY);
-    if (indexOfAnalyzersParam == -1) {
-      err.println(USAGE);
-      exitWithError();
+    if (useStdio) {
+      SonarLintLanguageServer.byStdio(analyzers, extraAnalyzers);
+      synchronized(this) {
```

Comment:
What is this used for?

Reply:
That was kind of a hack and I'm open for suggestions. It is necessary to keep reading from STDIN open and if I remove it the process will stop immediately.

Maybe, that piece of code should be done in `SonarLintLanguageServer` and the object should be notified when the server receives a shutdown request.

Reply:
I suggest you look at how it is implemented in other language servers based on LSP4J, like https://github.com/eclipse/eclipse.jdt.ls. I'm pretty sure they are using stdio as well.

Reply:
I found a solution for it by adding a `CountDownLatch` in `SonarLintLanguageServer` and waiting for it in `ServerMain`.